### PR TITLE
Add unavailabilities for an event to Availability Controller

### PR DESCRIPTION
**Pull request by MarioRodrigues10**
Make sure when you're testing this to also run https://github.com/coderdojobraga/shuriken/pull/174

### CODE REVIEW

File: lib/bokken_web/controllers/availability_controller.ex [hunk 1]
```diff
@@ -16,10 +16,11 @@ defmodule BokkenWeb.AvailabilityController do
 
   def index(conn, availability_params) do
     availabilities = Events.list_availabilities(availability_params, [:mentor])
+    unavailabilities = Events.list_unavailabilities(availability_params, [:mentor])
 
     conn
     |> put_status(:ok)
-    |> render("index.json", availabilities: availabilities)
+    |> render("index.json", availabilities: availabilities, unavailabilities: unavailabilities)
```
Comment (ruioliveira02) [inline lib/bokken_web/controllers/availability_controller.ex:23]: If you are not updating the view how does this work?

File: lib/bokken/events.ex [hunk 2]
```diff
@@ -726,6 +726,34 @@ defmodule Bokken.Events do
...
+    Availability
+    |> Repo.all()
+    |> Repo.preload(preloads)
+  end
+
+  def list_unavailabilities(%{"event_id" => event_id}, preloads \\ []) do
+    Availability
+    |> where([a], a.event_id == ^event_id and a.is_available == false)
+    |> Repo.all()
+    |> Repo.preload(preloads)
```
Comment (nelsonmestevao) [inline lib/bokken/events.ex]:
```suggestion
  def list_unavailabilities(%{"event_id" => event_id}, preloads \\ []) do
    Availability
    |> where([a], a.event_id == ^event_id and a.is_available == false)
    |> preload(preloads)
    |> Repo.all()
```

File: lib/bokken_web/controllers/availability_controller.ex (hunk 1 above)
Comment (MarioRodrigues10) [reply lib/bokken_web/controllers/availability_controller.ex:23]: I put all mentors related to an `event_id` in the same endpoint.

File: test/bokken/events_test.exs [hunk 3]
```diff
@@ -120,6 +120,23 @@ defmodule Bokken.EventsTest do
...
+
+    test "list_unavailabilities/1 returns all unavailabilities of the event" do
+      event = insert(:event)
+      availability = insert(:availability, %{event: event})
+
+      availabilities =
+        Events.list_availabilities(%{"event_id" => availability.event_id}, [:mentor])
+
+      assert hd(availabilities).id == availability.id
+    end
```
Comment (ruilopesm) [inline test/bokken/events_test.exs]: You should be testing unavailabilities, but you are using `list_availabilities` 😅

File: lib/bokken_web/controllers/availability_controller.ex (hunk 1 above)
Comment (ruilopesm) [reply lib/bokken_web/controllers/availability_controller.ex:23]: You are passing `unavailabilities` to the view, but not properly rendering it.

File: lib/bokken/events.ex [hunk 4]
```diff
@@ -721,7 +721,35 @@ defmodule Bokken.Events do
...
+
+      iex> list_unavailabilities(123)
+      [%Availability{is_available: false}, ...]
+
+  """
+  def list_unavailabilities(preloads) when is_list(preloads) do
+    Availability
+    |> Repo.all()
+    |> Repo.preload(preloads)
+  end
```
Comment (ruioliveira02) [inline lib/bokken/events.ex]:
```suggestion
  def list_unavailabilities(preloads) when is_list(preloads) do
    Availability
    |> where([a], not a.is_available)
    |> Repo.all()
    |> Repo.preload(preloads)
  end
```

File: lib/bokken/events.ex [hunk 5]
```diff
@@ -721,7 +721,35 @@ defmodule Bokken.Events do
...
+
+  def list_unavailabilities(%{"event_id" => event_id}, preloads \\ []) do
+    Availability
+    |> where([a], a.event_id == ^event_id and a.is_available == false)
```
Comment (ruioliveira02) [inline lib/bokken/events.ex]:
```suggestion
    |> where([a], a.event_id == ^event_id and not a.is_available)
```

File: lib/bokken_web/views/availability_view.ex [hunk 6]
```diff
@@ -35,6 +43,16 @@ defmodule BokkenWeb.AvailabilityView do
...
 
+  def render("unavailability.json", %{availability: availability, current_user: current_user}) do
+    %{
+      availability_id: availability.id,
+      is_available: availability.is_available,
+      notes: availability.notes
+    }
+    |> Map.merge(event(availability))
+    |> Map.merge(mentor(availability, current_user))
+  end
```
Comment (ruioliveira02) [inline lib/bokken_web/views/availability_view.ex]: This is exactly the same function as the one for `"availability.json"`.

File: test/support/factories/events_factory.ex [hunk 7]
```diff
@@ -20,7 +20,16 @@ defmodule Bokken.Factories.EventFactory do
...
+          notes: Lorem.sentence(0..10)
+        }
+      end
+
+      def unavailability_factory do
+        %Availability{
+          is_available: false,
           mentor: build(:mentor),
           event: build(:event),
           notes: Lorem.sentence(0..10)
```
Comment (ruioliveira02) [inline test/support/factories/events_factory.ex]: Exmachina allows you to specify attributes you want to have a certain value like so `insert(:availability, is_available: true)`.

So I don't think this distinction makes sense

File: test/bokken/events_test.exs [hunk 8]
```diff
@@ -120,6 +120,23 @@ defmodule Bokken.EventsTest do
...
+    test "list_unavailabilities/1 returns all unavailabilities of the event" do
+      event = insert(:event)
+      unavailability = insert(:unavailability, %{event: event})
+
+      unavailabilities =
+        Events.list_unavailabilities(%{"event_id" => unavailability.event_id}, [:mentor])
+
+      assert hd(unavailabilities).id == unavailability.id
+    end
+
```
Comment (ruioliveira02) [inline test/bokken/events_test.exs]: You should add a test where you insert an availability, just to make sure you are not returning availabilities

Comment (MarioRodrigues10) [reply test/bokken/events_test.exs]: That was already done in previous tests

File: lib/bokken_web/views/availability_view.ex (hunk 6 above)
Comment (MarioRodrigues10) [reply lib/bokken_web/views/availability_view.ex]: Yes it's but i need a different json to load unavailabilities

File: lib/bokken/events.ex (hunk 4 above)
Comment (MarioRodrigues10) [reply lib/bokken/events.ex]: Should i also change `list_availabilities` ?

File: lib/bokken_web/views/availability_view.ex (hunk 6 above)
Comment (ruioliveira02) [reply lib/bokken_web/views/availability_view.ex]: Why?

**Review by ruioliveira02 — APPROVED** [review]

**Pull request merged**